File: leetcode/string/wildcard_matching.go
package main

import "fmt"

func isMatch(s string, p string) bool {
	s = " " + s
	p = " " + p
	rows, cols := len(s), len(p)
	dp := make([][]bool, rows)
	for i := range dp {
		dp[i] = make([]bool, cols)
	}
	dp[0][0] = true
	for col := 1; col < cols; col++ {
		if p[col] == 42 {
			dp[0][col] = dp[0][col-1]
		}
	}

	for row := 1; row < rows; row++ {
		for col := 1; col < cols; col++ {
			if s[row] == p[col] || p[col] == 63 {
				dp[row][col] = dp[row-1][col-1]
			} else if p[col] == 42 {
				dp[row][col] = dp[row-1][col] || dp[row][col-1]
			}
		}
	}
	return dp[rows-1][cols-1]
}

func main() {
	fmt.Println(isMatch("adceb", "*a*b"))
}

File: leetcode/string/wildcard_matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        s = " " + s
        p = " " + p
        rows, cols = len(s), len(p)
        dp = [[False]*cols for _ in range(rows)]
        dp[0][0] = True
        for col in range(1, cols):
            if p[col] == "*":
                dp[0][col] = dp[0][col-1]
        for row in range(1, rows):
            for col in range(1, cols):
                if s[row] == p [col] or p[col] == '?':
                    dp[row][col] = dp[row-1][col-1]
                elif p[col] == '*':
                    dp[row][col] = dp[row][col-1] or dp[row-1][col]
        return dp[-1][-1]

if __name__ == "__main__":
    print(Solution().isMatch("adceb", "*a*b"))